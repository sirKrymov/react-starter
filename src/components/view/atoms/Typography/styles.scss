// @NOTE: 1rem = 16px by default
@import '../../../../styles/mixins';

.typography {
  &--display-initial {
    display: initial;
  }

  &--display-block {
    display: block;
  }

  &--display-inline {
    display: inline;
  }

  &--variant-inherit {
    font-weight: inherit;
    line-height: inherit;
    font-size: inherit;
  }

  &--variant-subtitle1 {
    font-weight: 500;
    line-height: 1.4;
    font-size: 1.5rem;
  }

  &--variant-subtitle2 {
    font-weight: 500;
    line-height: 1.4;
    font-size: 1.125rem;
  }

  &--variant-subtitle3 {
    font-weight: 500;
    line-height: 1.4;
    font-size: 1rem;
  }

  &--variant-caption {
    line-height: 1.3;
    font-weight: 400;
    font-size: 0.625rem;
  }

  &--variant-body1 {
    font-weight: 400;
    line-height: 1.5;
    font-size: 1.125rem;
  }

  &--variant-body2 {
    font-weight: 400;
    line-height: 1.5;
    font-size: 1rem;
  }

  &--variant-body3 {
    font-weight: 400;
    line-height: 1.5;
    font-size: 0.875rem;
  }

  &--variant-body4 {
    font-weight: 400;
    line-height: 1.5;
    font-size: 12px;
  }

  &--variant-h1 {
    font-weight: 700;
    line-height: 1.4;
    font-size: 2.5rem;
  }

  &--variant-h2 {
    font-weight: 700;
    line-height: 1.4;
    font-size: 2rem;
  }

  &--variant-h3 {
    font-weight: 700;
    line-height: 1.4;
    font-size: 1.5rem;
  }

  &--variant-h4 {
    font-weight: 700;
    line-height: 1.4;
    font-size: 1.125rem;
  }

  &--variant-h5 {
    font-weight: 700;
    line-height: 1.4;
    font-size: 1rem;
  }

  &--variant-h6 {
    font-weight: 700;
    line-height: 1.4;
    font-size: 0.875rem;
  }

  &--color-textSecondary {
    color: var(--textSecondary);
  }

  &--color-textPrimary {
    color: var(--textPrimary);
  }

  &--color-textThirdy {
    color: var(--textThirdy);
  }

  &--color-textLight {
    color: var(--textLight);
  }

  &--color-textDark {
    color: var(--textDark);
  }

  &--color-secondary {
    color: var(--secondary);
  }

  &--color-primary {
    color: var(--primary);
  }

  &--color-thirdy {
    color: var(--thirdy);
  }

  &--color-initial {
    color: initial;
  }

  &--color-inherit {
    color: inherit;
  }

  &--color-valid {
    color: var(--valid);
  }

  &--color-error {
    color: var(--error);
  }

  &--align-left {
    text-align: left;
  }

  &--align-center {
    text-align: center;
  }

  &--align-right {
    text-align: right;
  }

  &--align-justify {
    text-align: justify;
  }

  &--font-roboto {
    font-family: 'Roboto';
  }

  &--uppercase {
    text-transform: uppercase;
  }

  &--nowrap {
    max-width: 100%;
    display: block;
    width: 100%;
    @include text-ellipsis;
  }
}
