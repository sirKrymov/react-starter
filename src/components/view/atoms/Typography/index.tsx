// @SETUP: Please don't forget to remove this component or current comment if component is used in project
import React, { ReactNode, ReactElement } from 'react';

import cn from 'classnames';

import { capitalize } from 'utils';

import './styles.scss';

interface IProps {
  isCapitalize?: boolean;
  component?: 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' | 'div' | 'p' | 'span';
  paragraph?: boolean;
  uppercase?: boolean;
  className?: string;
  children: ReactNode | string | number;
  variant?:
    | 'subtitle1'
    | 'subtitle2'
    | 'subtitle3'
    | 'inherit'
    | 'caption'
    | 'body1'
    | 'body2'
    | 'body3'
    | 'body4'
    | 'h1'
    | 'h2'
    | 'h3'
    | 'h4'
    | 'h5'
    | 'h6';
  display?: 'initial' | 'block' | 'inline';
  noWrap?: boolean;
  color?:
    | 'textSecondary'
    | 'textPrimary'
    | 'textThirdy'
    | 'textLight'
    | 'secondary'
    | 'textDark'
    | 'primary'
    | 'initial'
    | 'inherit'
    | 'thirdy'
    | 'error'
    | 'valid';
  align?: 'inherit' | 'left' | 'center' | 'right' | 'justify';
  font?: 'roboto';
}

export function Typography({
  isCapitalize,
  component,
  paragraph,
  uppercase,
  className,
  children,
  variant,
  display = 'initial',
  noWrap,
  color = 'textPrimary',
  align = 'left',
  font = 'roboto'
}: IProps): ReactElement {
  const tagMapping = {
    subtitle1: 'h6',
    subtitle2: 'h6',
    subtitle3: 'h6',
    inherit: 'p',
    body1: 'p',
    body2: 'p',
    body3: 'p',
    body4: 'p',
    h1: 'h1',
    h2: 'h2',
    h3: 'h3',
    h4: 'h4',
    h5: 'h5',
    h6: 'h6'
  };
  const Tag =
    component || (paragraph ? 'p' : variant && tagMapping[variant]) || 'span';

  if (isCapitalize && typeof children === 'string')
    children = capitalize(children);

  const typographyClass = cn(
    {
      typography: true,
      [`typography--display-${display}`]: display,
      [`typography--variant-${variant}`]: variant,
      [`typography--color-${color}`]: color,
      [`typography--align-${align}`]: align,
      [`typography--font-${font}`]: font,
      [`typography--uppercase`]: uppercase,
      [`typography--nowrap`]: noWrap
    },
    className
  );

  return <Tag className={typographyClass}>{children}</Tag>;
}
