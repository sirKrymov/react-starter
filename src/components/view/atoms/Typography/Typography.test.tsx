import React from 'react';

import { shallow, ShallowWrapper } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import { Typography } from '../Typography';

describe('<Typography> atom component', () => {
  let shallowComponent: ShallowWrapper;

  const createComponent = (props?: Record<string, any>) => {
    shallowComponent = shallow(
      <Typography {...props}>text for test</Typography>
    );
  };

  describe('WHEN the component loads', () => {
    beforeEach(() => {
      createComponent();
    });

    it('THEN it should render without crashing', () => {
      expect(shallowComponent.html()).not.toHaveLength(0);
    });

    it('THEN it should match snapshot', () => {
      expect(shallowToJson(shallowComponent)).toMatchSnapshot();
    });
  });

  describe('WHEN the component have different props', () => {
    it('WHEN component have prop component="div"', () => {
      createComponent({ component: 'div' });

      expect(shallowComponent.find('div.typography')).toHaveLength(1);
    });

    it('WHEN component have paragraph prop', () => {
      createComponent({ paragraph: true });

      expect(shallowComponent.find('p.typography')).toHaveLength(1);
    });

    it('WHEN component have prop variant="h1"', () => {
      createComponent({ variant: 'h1' });

      expect(shallowComponent.find('h1.typography')).toHaveLength(1);
    });

    it('WHEN component have prop isCapitalize', () => {
      createComponent({ isCapitalize: true });

      expect(shallowComponent.prop('children')).toEqual('Text for test');
    });
  });
});
