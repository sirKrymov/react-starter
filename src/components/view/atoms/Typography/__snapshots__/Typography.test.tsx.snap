// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Typography> atom component WHEN the component loads THEN it should match snapshot 1`] = `
<span
  className="typography typography--display-initial typography--color-textPrimary typography--align-left typography--font-roboto"
>
  text for test
</span>
`;
